### PR TITLE
Add hot code reload

### PR DESCRIPTION
**Pull request by iusildra**
Address this [feature request](https://github.com/scalameta/metals-feature-requests/issues/336), but for now, it should only works with sbt, I'll look to make it works with bloop too later. I adapted the implementation from [vscode-java-debug](https://github.com/microsoft/vscode-java-debug) to work with metals

The implementation on the debugger side is in [this pr](https://github.com/scalacenter/scala-debug-adapter/pull/553) (not published yet)

There are a few inconsistencies left in my implementation on the VS Code extension:

- [x] If the setting `scalaHotReloadOn` is not ticked, the button still appear (we need to tick and untick to make it disappear)
- [x] If HCR is not supported by the debugger, it show a warning message "no changed classes found", which is not relevant

### CODE REVIEW

**Comment by iusildra** [comment]
It will need both upcoming versions of the debugger (for the implementation) and metals (for the sync with compilation) to be fully functional

Also, before merging I'll need to change the metals server's version before which HCR would not work to give proper information

File: packages/metals-vscode/src/hotCodeReplace.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  const debugSession: DebugSession | undefined =
+    vscode.debug.activeDebugSession;
+  if (!debugSession) {
+    return;
+  }
+
+  const serverVersion =
+    vscode.workspace.getConfiguration("metals").get<string>("serverVersion") ??
+    "";
+  if (serverVersion < "1.0.1") {
```
Comment (tgodzik) [inline packages/metals-vscode/src/hotCodeReplace.ts:44]: I think we could just make sure we update DAP in Metals and in Bloop first and once that is available we can merge this PR and update the version to the most recent snapshot at the time. If anyone switches to an older version of Metals it's fine if we fail.

**Review by tgodzik — CHANGES_REQUESTED** [review]
Thanks for the great work! It will be amazing for the users.

Let's wait until we update scala debug adapter properly in Bloop and sbt

File: packages/metals-vscode/package.json [hunk 2]
```diff
@@ -218,6 +218,11 @@
     "configuration": {
       "title": "Metals",
       "properties": {
+        "metals.debug.settings.hotCodeReplace": {
+          "type": "boolean",
+          "default": false,
+          "markdownDescription": "Allow Hot Code Replace (HCR) while debugging. Only available since Metals 1.0.X"
```
Comment (tgodzik) [inline packages/metals-vscode/package.json:224]:
```suggestion
          "markdownDescription": "Allow Hot Code Replace (HCR) while debugging."
```

File: packages/metals-vscode/package.json [hunk 3]
```diff
@@ -1085,7 +1105,8 @@
     "metals-languageclient": "file:../metals-languageclient",
     "promisify-child-process": "4.1.1",
     "semver": "^7.5.2",
-    "vscode-languageclient": "8.1.0"
+    "vscode-languageclient": "8.1.0",
+    "vscode-extension-telemetry-wrapper": "^0.13.3"
```
Comment (tgodzik) [inline packages/metals-vscode/package.json:1109]: Is this needed?

File: packages/metals-vscode/package.json [hunk 4]
```diff
@@ -218,6 +218,11 @@
     "configuration": {
       "title": "Metals",
       "properties": {
+        "metals.debug.settings.hotCodeReplace": {
```
Comment (tgodzik) [inline packages/metals-vscode/package.json:221]:
```suggestion
        "metals.hotCodeReplace": {
```
maybe is enough?

File: packages/metals-vscode/src/hotCodeReplace.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  const response = await redefineRequest;
+
+  if (response?.errorMessage) {
+    vscode.window.showErrorMessage(response.errorMessage);
+    return;
+  }
+
+  const NO_HCR = "Disable HCR";
+
+  if (!response?.changedClasses?.length) {
```
Comment (tgodzik) [inline packages/metals-vscode/src/hotCodeReplace.ts:71]: Is this needed, why would we want to disable HCR here?

File: packages/metals-vscode/src/hotCodeReplace.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    const res = await vscode.window.showWarningMessage(
+      "No classes were reloaded",
+      "Ok",
+      NO_HCR
+    );
+
+    if (res === NO_HCR) {
+      vscode.workspace
+        .getConfiguration("metals.debug.settings")
+        .update("hotCodeReplace", false, vscode.ConfigurationTarget.Workspace);
```
Comment (tgodzik) [inline packages/metals-vscode/src/hotCodeReplace.ts:81]: Probably would be good to put `"hotCodeReplace"` and `"scalaHotReloadOn"` into const values

File: .vscode/settings.json [hunk 7]
```diff
@@ -10,5 +10,25 @@
   "typescript.tsdk": "node_modules/typescript/lib",
   "files.watcherExclude": {
     "**/target": true
-  }
+  },
+  "workbench.colorCustomizations": {
```
[the remark this answers is not in the document]
Comment (iusildra) [reply .vscode/settings.json:14]: oups, yeah will do

File: packages/metals-vscode/src/hotCodeReplace.ts (hunk 5 above)
Comment (iusildra) [reply packages/metals-vscode/src/hotCodeReplace.ts:71]: Forgot to remove it. It was when I wasn't sure of how to handle the case where HCR is not supported

File: packages/metals-vscode/package.json (hunk 3 above)
Comment (iusildra) [reply packages/metals-vscode/package.json:1109]: Good point, `registerCommand` would do the same

**Comment by adpi2** [comment]
@tgodzik  this PR is ready to be reviewed and merged.

It adds the lightning button in the debug tool bar to trigger a hot code reload.

File: packages/metals-vscode/package.json [hunk 8]
```diff
@@ -343,6 +343,11 @@
           "default": false,
           "markdownDescription": "When this option is enabled, when user pastes any snippet into a Scala file, Metals will try to adjust the indentation to that of the current cursor."
         },
+        "metals.debug.hotCodeReplace": {
```
Comment (tgodzik) [inline packages/metals-vscode/package.json:346]: MAybe we don't need a setting actually? If it doesn't work then the button would not work and that's fine as long it doesn't break anything else

Comment (adpi2) [reply packages/metals-vscode/package.json:346]: Okay, I'll remove the setting

Comment (adpi2) [reply packages/metals-vscode/package.json:346]: I don't think we need

**Review by tgodzik — APPROVED** [review]
LGTM!

**Pull request merged**